File: chimera/src/token.rs
use phf::{phf_map, Map};

pub type TokenType = &'static str;

#[derive(Default, Debug)]
pub struct Token {
   pub token_type: TokenType,
   pub literal: String 
}

pub static ILLEGAL: &str = "ILLEGAL";
pub static EOF: &str = "EOF";

// Identifiers & Literals
pub static IDENT: &str = "IDENT";
pub static INT: &str = "INT";

// Operators
pub static ASSIGN: &str = "=";
pub static PLUS: &str = "+";
pub static MINUS: &str = "-";
pub static BANG: &str = "!";
pub static ASTERISK: &str = "*";
pub static SLASH: &str = "/";

pub static LT: &str = "<";
pub static GT: &str = ">";

pub static EQ: &str = "==";
pub static NQ: &str = "!=";

// Delimeters
pub static COMMA: &str = ",";
pub static SEMICOLON: &str = ";";

pub static LPAREN: &str = "(";
pub static RPAREN: &str = ")";
pub static LBRACE: &str = " {";
pub static RBRACE: &str = "}";

// Keywords
pub static FUNCTION: &str = "FUNCTION";
pub static LET: &str = "LET";
pub static TRUE: &str = "TRUE";
pub static FALSE: &str = "FALSE";
pub static IF: &str = "IF";
pub static ELSE: &str = "ELSE";
pub static RETURN: &str = "RETURN";

static KEYWORDS: Map<&str, TokenType> = phf_map! {
    "fn" => FUNCTION,
    "let" => LET,
    "true" => TRUE,
    "false" => FALSE,
    "if" => IF,
    "else" => ELSE,
    "return" => RETURN,
};

pub fn lookup_ident(identifier: &str ) -> TokenType {
    if KEYWORDS.contains_key(identifier) {
        return KEYWORDS.get(identifier).unwrap()
    };
    IDENT
}

File: c-interpreter/c-interpreter/src/token.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include "../include/token.h"

const char ILLEGAL[] = "ILLEGAL";
const char EOF_IDENT[] = "EOF";

const char IDENT[] = "IDENT";
const char INTEGER[] = "INT";
const char ASSIGN[] = "=";
const char PLUS[] = "+";
const char MINUS[] = "-";
const char BANG[] = "!";
const char ASTERIX[] = "*";
const char SLASH[] = "/";

const char LT[] = "<";
const char GT[] = ">";

const char EQ[] = "==";
const char NQ[] = "!=";

const char COMMA[] = ",";
const char SEMICOLON[] = ";";

const char LPAREN[] = "(";
const char RPAREN[] = ")";
const char LBRACE[] = "{";
const char RBRACE[] = "}";

const char FUNCTION[] = "FUNCTION";
const char LET[] = "LET";
const char TRUE_KEYWORD[] = "TRUE";
const char FALSE_KEYWORD[] = "FALSE";
const char IF_KEYWORD[] = "IF";
const char ELSE_KEYWORD[] = "ELSE";
const char RETURN_KEYWORD[] = "RETURN";

TokenTypeMap keywords[] = {
    {"fn", FUNCTION},
    {"let", LET},
    {"true", TRUE_KEYWORD},
    {"false", FALSE_KEYWORD},
    {"if", IF_KEYWORD},
    {"else", ELSE_KEYWORD},
    {"return", RETURN_KEYWORD},
};

P_TokenType lookupIdent(char *ident) {
    for (int i = 0; i < MAX_SIZE; i++) {
        if (strcmp(keywords[i].key, ident) == 0) {
            return keywords[i].value;
        }
    }
    return IDENT;
}

Token *new_Token(P_TokenType tokenType, char *literal) {
    Token *t = malloc(sizeof(Token));
    
    if (t == NULL) {
        printf("Error: Could not allocate memory.\n");
        return NULL;
    }

    char *temp_literal = strdup(literal);
    if (temp_literal == NULL) {
        free(t);
        return NULL;
    }

    t->type = tokenType;
    t->literal = temp_literal;

    return t;
}

void free_Token(Token *t) {
    free(t->literal);
    free(t);
}
